comp/core/workloadmeta/collectors/internal/process/process_collector.go

```diff
@@ -0,0 +1,220 @@
+	defer cancel()
+
+	for {
+		select {
+		case evt, ok := <-containerEvt:
+			if !ok {
+				log.Infof("The %s collector has stopped, workloadmeta channel is closed", collectorID)
+				return
+			}
+			c.handleContainerEvent(evt)
+		case <-collectionTicker.C:
+			err := c.processData.Fetch()
+			if err != nil {
+				log.Error("Error fetching process data:", err)
+			}
+		case <-ctx.Done():
+			log.Infof("The %s collector has stopped", collectorID)
+			return
+		}
+	}
+}
+
+func (c *collector) stream(ctx context.Context) {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	health := health.RegisterLiveness(componentName)
+	for {
+		select {
+		case <-health.C:
+
+		case diff := <-c.processDiffCh:
+			log.Debugf("Received process diff with %d creations and %d deletions", len(diff.Creation), len(diff.Deletion))
+			events := transform(diff)
+			c.store.Notify(events)
+
+		case <-ctx.Done():
+			err := health.Deregister()
+			if err != nil {
+				log.Warnf("error de-registering health check: %s", err)
+			}
+			return
+		}
+	}
+}
+
+func (c *collector) Pull(_ context.Context) error {
+	return nil
+}
+
+func (c *collector) GetID() string {
+	return c.id
+}
+
+func (c *collector) GetTargetCatalog() workloadmeta.AgentType {
+	return c.catalog
+}
+
+// transform converts a ProcessCacheDiff into a list of CollectorEvents.
+// The type of event is based whether a process was created or deleted since the last diff.
+func transform(diff *processwlm.ProcessCacheDiff) []workloadmeta.CollectorEvent {
+	events := make([]workloadmeta.CollectorEvent, 0, len(diff.Creation)+len(diff.Deletion))
+
+	for _, creation := range diff.Creation {
+		events = append(events, workloadmeta.CollectorEvent{
+			Type: workloadmeta.EventTypeSet,
+			Entity: &workloadmeta.Process{
+				EntityID: workloadmeta.EntityID{
+					Kind: workloadmeta.KindProcess,
+					ID:   strconv.Itoa(int(creation.Pid)),
+				},
+				ContainerID:  creation.ContainerId,
+				NsPid:        creation.NsPid,
+				CreationTime: time.UnixMilli(creation.CreationTime),
+				Language:     creation.Language,
+			},
+			Source: workloadmeta.SourceLocalProcessCollector,
+		})
+	}
+
+	for _, deletion := range diff.Deletion {
+		events = append(events, workloadmeta.CollectorEvent{
+			Type: workloadmeta.EventTypeUnset,
+			Entity: &workloadmeta.Process{
+				EntityID: workloadmeta.EntityID{
+					Kind: workloadmeta.KindProcess,
+					ID:   strconv.Itoa(int(deletion.Pid)),
+				},
+			},
+			Source: workloadmeta.SourceLocalProcessCollector,
+		})
+	}
+
+	return events
+}
+
+func (c *collector) handleContainerEvent(evt workloadmeta.EventBundle) {
+	defer evt.Acknowledge()
+
+	for _, evt := range evt.Events {
+		ent := evt.Entity.(*workloadmeta.Container)
+		switch evt.Type {
+		case workloadmeta.EventTypeSet:
+			// Should be safe, even on windows because PID 0 is the idle process and therefore must always belong to the host
+			if ent.PID != 0 {
+				c.pidToCid[ent.PID] = ent.ID
+			}
```

Comment:
Here `ent.PID` is the container's main process ID in the OS, and `ent.ID` is the container ID in the container runtime.

Hence, with this, `pidToCid` will only map main container process PID to the container ID. Any process created within the container will not be included in this map, which is an issue for APM SSI.

IMHO, building this map based on received container events is not sufficient because container entity events don't contain complete data (non-main processes of containers are missing).  

What I think would be better is to use the shared [container provider](https://github.com/DataDog/datadog-agent/blob/fd3298cf940ad6df3e3f45ef77f2fd173b2ed649/pkg/process/util/containers/containers.go#L59-L62) to construct this mapping. The container provider exposes a GetContainers method which returns, among other data, the complete pidToCid mapping. But this comes at a cost because this function not only calculates pidToCid mapping, but also container stats and other stuff which consumes memory and cpu and are of no use here. For this reason, and supposing there is no ready-to-use function that calculates the pidToCid, I'd suggest extending the existing interface with a `GetPidToCid` method that only returns the required map. The logic of this method can be copy-pasted from the `GetContainers` method while removing the logic that computes metrics/stats for containers that we don't need here.

Reply:
Another alternative that can be possible is the following: 

Instead of maintaining a mapping from pid to cid in the wlm extractor, you might be able to rely on the shared metacollector to get the container id for a specific process id on the fly with [GetContainerIdForPid](https://github.com/DataDog/datadog-agent/blob/9ae6e7fb7671e8f44b3f92bbeced6ee62ae6bb12/pkg/util/containers/metrics/provider/metacollector.go#L58-L75) method.

Reply:
I used the first approach. Let me know what you think